### PR TITLE
Issue #465: adds caching for metadata around PackageObjectFactory

### PR DESCRIPTION
**Pull request by rnveach**
Issue #465

I did a no harm test by having a test run `new CheckstyleMetadata(repository).createRulesWithMetadata();`. It loaded 40 checks with no issues.

### CODE REVIEW

**Review by muhlba91 — APPROVED** [review]

File: src/main/java/org/sonar/plugins/checkstyle/metadata/CheckUtil.java [hunk 1]
```diff
@@ -60,11 +62,19 @@ else if (AbstractCheck.class.isAssignableFrom(checkResult.getClass())) {
...
     private static AbstractCheck getCheck(String checkName) {
-        final ClassLoader classLoader = CheckstyleMetadata.class.getClassLoader();
+        if (packageObjectFactory == null) {
+            try {
+                final ClassLoader classLoader = CheckstyleMetadata.class.getClassLoader();
+                final Set<String> packageNames = PackageNamesLoader.getPackageNames(classLoader);
+                packageObjectFactory = new PackageObjectFactory(packageNames, classLoader);
+            }
+            catch (CheckstyleException ex) {
+                throw new IllegalStateException(ex);
```
Comment (romani) [inline src/main/java/org/sonar/plugins/checkstyle/metadata/CheckUtil.java:72]: During wrapping of exception we should provide context of execution in this case it is checkname

**Review by romani — CHANGES_REQUESTED** [review]
Items

File: src/main/java/org/sonar/plugins/checkstyle/metadata/CheckUtil.java (hunk 1 above)
Comment (rnveach) [reply src/main/java/org/sonar/plugins/checkstyle/metadata/CheckUtil.java:72]: There is no extra "context" to provide.

The only parameter of the method, `checkName`, is not used in this try/catch so it has no basis on the exception. The only thing to know is what method did the call and caused the exception, and that is already part of the stack trace.

Any exception that does occur, all information on the issue will be from the original exception.

**Comment by rnveach** [comment]
@romani ping

File: src/main/java/org/sonar/plugins/checkstyle/metadata/CheckUtil.java (hunk 1 above)
Comment (romani) [reply src/main/java/org/sonar/plugins/checkstyle/metadata/CheckUtil.java:72]: IllegalStateException("exception happened during loading of " + checkName, ex);

File: src/main/java/org/sonar/plugins/checkstyle/metadata/CheckUtil.java [hunk 2]
```diff
@@ -34,6 +34,8 @@
 import com.puppycrawl.tools.checkstyle.utils.TokenUtil;
 
 public final class CheckUtil {
+    private static PackageObjectFactory packageObjectFactory;
```
Comment (romani) [inline src/main/java/org/sonar/plugins/checkstyle/metadata/CheckUtil.java:37]: Util class should stay stateless.
Let's move this factory to class that calls getCheck

**Review by romani — CHANGES_REQUESTED** [review]
Items

File: src/main/java/org/sonar/plugins/checkstyle/metadata/CheckUtil.java (hunk 1 above)
Comment (rnveach) [reply src/main/java/org/sonar/plugins/checkstyle/metadata/CheckUtil.java]: Done

File: src/main/java/org/sonar/plugins/checkstyle/metadata/CheckUtil.java (hunk 2 above)
Comment (rnveach) [reply src/main/java/org/sonar/plugins/checkstyle/metadata/CheckUtil.java]: Done.

**Review by romani — APPROVED** [review]
Ok to merge

**Pull request merged**